packages/eslint-plugin-react-hooks/src/RulesOfHooks.js

```diff
@@ -112,6 +135,41 @@ export default {
   create(context) {
     const codePathReactHooksMapStack = [];
     const codePathSegmentStack = [];
+    const useEventViolations = new Set();
+
+    /**
+     * For a given AST node, traverse into it and add all useEvent definitions. We can do this in
+     * non-Program nodes because we can rely on the assumption that useEvent functions can only be
+     * declared within a component or hook.
+     *
+     * @param {Node} node
+     */
+    function addAllUseEventViolations(node) {
+      traverse(context, node, childNode => {
```

Comment:
It looks like we're still adding two traversals in this iteration. I'm sorry to keep pushing on this but I want to understand if we truly can't avoid this. Generally we should try to do as much work as possible within the traversal ESLint already provides us. The more work our lint rules are doing, the more they slow down the builds, the more likelihood people will turn them off in CI, etc.

I was imagining that it should be possible to do something like this:

- When we step into what looks like a React function, we scan it shallowly for top-level declarations that are result of `useEvent` calls. Don't traverse deeply — I think it should be enough to go over the top-level assignment statements. This gives us a list of bindings that represent event functions.
- When we step into an arrow function, we check if it's a `useEffect` first argument. If it is, we set some state like "we're in an effect now!".
- When we step out of an arrow function, we check if "we're in an effect now". If we are, and this is the same function we previously stepped into, then reset the "we're in an effect now" state.
- When we meet an identifier, we check if the binding resolves to one of the event bindings. If yes, and we're not "in an effect now", and also it's not inside a call expression, then we flag a violation.
- When we step out of a React function, we clean the map of bindings.

Would this not work?

Reply:
I think that's fair to call out, not handling the case where the useEvent function is referenced prior to its definition (like in the nested component case) does simplify the code a lot and means we can get away with the shallower traversals you're describing in your example.

Reply:
I think I’m missing why that case wouldn’t work? I imagine you’d shallowly scan the entire body of the function once you enter it — without waiting to enter the declarator. So that would collect all definitions first regardless of the ordering. And then the rest of the logic runs with existing knowledge of all useEvent declarations, so it should work whether it’s nested or not. What case am I missing?

Reply:
Referencing before definition is important to work regardless of whether components are nested. eg this needs to work:

```js
useEffect(() => { onFoo() })
const onFoo = useEvent(…)
```

Reply:
@gaearon Ah whoops, you're right! I got mixed up with something else I had done earlier